File: backend/app/schemas/product.py
from typing import Any, Dict, List, Literal, Optional
from pydantic import BaseModel, Field

ProductFieldStatus = Literal["detected", "not_printed", "not_visible", "uncertain", "not_applicable"]
ProductFieldSource = Literal["ocr", "vision", "merged", "user", "none"]

# Core Legal Metrology compliance fields (16 primary fields)
CORE_LEGAL_METROLOGY_FIELDS = (
    "commodity_name",
    "generic_name",
    "net_quantity",
    "quantity_unit",
    "manufacturer_name",
    "manufacturer_address",
    "packer_name",
    "packer_address",
    "importer_name",
    "importer_address",
    "country_of_origin",
    "mrp",
    "mrp_tax_inclusive",
    "unit_sale_price",
    "manufacturing_date",
    "best_before_date",
    "consumer_care_details",
)

# Canonical 26 fields across pipeline for backward compatibility
PRODUCT_FIELDS = (
    "brand_or_commodity_name",
    "generic_name",
    "net_quantity",
    "quantity_unit",
    "manufacturer_name",
    "manufacturer_address",
    "packer_name",
    "packer_address",
    "importer_name",
    "importer_address",
    "marketer_name",
    "marketer_address",
    "mrp",
    "mrp_tax_inclusive",
    "unit_sale_price",
    "packing_date",
    "manufacturing_date",
    "expiry_date",
    "batch_number",
    "customer_care_name",
    "customer_care_phone",
    "toll_free_number",
    "customer_care_email",
    "country_of_origin",
    "vegetarian_mark",
    "non_vegetarian_mark",
    "fssai_number",
    "certifications",
)

# Fields whose disagreement must never be silently resolved.
CONFLICT_SENSITIVE_FIELDS = (
    "mrp",
    "net_quantity",
    "packing_date",
    "manufacturing_date",
    "expiry_date",
    "fssai_number",
)


class ProductField(BaseModel):
    """Per-field extraction outcome.

    value      - printed/extracted value, or null when absent/unreadable
    status     - detected | not_printed | not_visible | uncertain | not_applicable
    confidence - 0-100, real confidence from the producing source
    source     - ocr | vision | user | none | merged
    normalized - optional standardized string representation
    conflicts  - optional debugging metadata when OCR and vision disagree
    """

    value: Optional[str] = None
    status: ProductFieldStatus = "not_visible"
    confidence: float = Field(default=0.0, ge=0.0, le=100.0)
    source: ProductFieldSource = "none"
    normalized: Optional[str] = None
    conflicts: Optional[List[Dict[str, Any]]] = None


def field(
    value: Optional[str] = None,
    status: str = "not_visible",
    confidence: float = 0.0,
    source: str = "none",
    normalized: Optional[str] = None,
    conflicts: Optional[List[Dict[str, Any]]] = None,
) -> ProductField:
    return ProductField(
        value=value,
        status=status,  # type: ignore[arg-type]
        confidence=confidence,
        source=source,  # type: ignore[arg-type]
        normalized=normalized,
        conflicts=conflicts,
    )


class OtherDetectedInformation(BaseModel):
    """Product details detected by OCR/Vision that are NOT Legal Metrology compliance rules.
    
    They do NOT affect Legal Metrology compliance scores or create violations.
    """
    brand_name: Optional[str] = None
    marketer_name: Optional[str] = None
    marketer_address: Optional[str] = None
    batch_number: Optional[str] = None
    fssai_number: Optional[str] = None
    vegetarian_mark: Optional[str] = None
    non_vegetarian_mark: Optional[str] = None
    nutrition_info: Dict[str, str] = Field(default_factory=dict)
    ingredients: Optional[str] = None
    certifications: Optional[str] = None


class ProductInformation(BaseModel):
    """Canonical structured product-information schema with Legal Metrology fields."""

    # Core Legal Metrology
    brand_or_commodity_name: ProductField = field()
    generic_name: ProductField = field()
    net_quantity: ProductField = field()
    quantity_unit: ProductField = field()
    manufacturer_name: ProductField = field()
    manufacturer_address: ProductField = field()
    packer_name: ProductField = field()
    packer_address: ProductField = field()
    importer_name: ProductField = field()
    importer_address: ProductField = field()
    country_of_origin: ProductField = field()
    mrp: ProductField = field()
    mrp_tax_inclusive: ProductField = field()
    unit_sale_price: ProductField = field()
    manufacturing_date: ProductField = field()
    packing_date: ProductField = field()
    expiry_date: ProductField = field()
    customer_care_name: ProductField = field()
    customer_care_phone: ProductField = field()
    toll_free_number: ProductField = field()
    customer_care_email: ProductField = field()

    # Legacy / Non-LM attributes maintained for compatibility
    marketer_name: ProductField = field()
    marketer_address: ProductField = field()
    batch_number: ProductField = field()
    vegetarian_mark: ProductField = field()
    non_vegetarian_mark: ProductField = field()
    fssai_number: ProductField = field()
    certifications: ProductField = field()

    # Separate container for non-LM information
    other_detected_information: OtherDetectedInformation = Field(default_factory=OtherDetectedInformation)

    def as_dict(self) -> Dict[str, ProductField]:
        return {name: getattr(self, name) for name in PRODUCT_FIELDS if hasattr(self, name)}

    def model_dump_canonical(self) -> Dict[str, Dict[str, Any]]:
        return {name: getattr(self, name).model_dump() for name in PRODUCT_FIELDS if hasattr(self, name)}


class ProductExtraction(BaseModel):
    """Structured output schema requested from the vision LLM (Gemini)."""

    image_quality_flag: bool = False
    raw_text: Optional[str] = None
    brand_or_commodity_name: ProductField = Field(default_factory=field)
    generic_name: ProductField = Field(default_factory=field)
    net_quantity: ProductField = Field(default_factory=field)
    quantity_unit: ProductField = Field(default_factory=field)
    manufacturer_name: ProductField = Field(default_factory=field)
    manufacturer_address: ProductField = Field(default_factory=field)
    packer_name: ProductField = Field(default_factory=field)
    packer_address: ProductField = Field(default_factory=field)
    importer_name: ProductField = Field(default_factory=field)
    importer_address: ProductField = Field(default_factory=field)
    marketer_name: ProductField = Field(default_factory=field)
    marketer_address: ProductField = Field(default_factory=field)
    mrp: ProductField = Field(default_factory=field)
    mrp_tax_inclusive: ProductField = Field(default_factory=field)
    unit_sale_price: ProductField = Field(default_factory=field)
    packing_date: ProductField = Field(default_factory=field)
    manufacturing_date: ProductField = Field(default_factory=field)
    expiry_date: ProductField = Field(default_factory=field)
    batch_number: ProductField = Field(default_factory=field)
    customer_care_name: ProductField = Field(default_factory=field)
    customer_care_phone: ProductField = Field(default_factory=field)
    toll_free_number: ProductField = Field(default_factory=field)
    customer_care_email: ProductField = Field(default_factory=field)
    country_of_origin: ProductField = Field(default_factory=field)
    vegetarian_mark: ProductField = Field(default_factory=field)
    non_vegetarian_mark: ProductField = Field(default_factory=field)
    fssai_number: ProductField = Field(default_factory=field)
    certifications: ProductField = Field(default_factory=field)

File: frontend/types/product.ts
// Canonical product-information schema (mirrors backend/app/schemas/product.py).
// Single source of truth for the dynamic results grid and any future edit UI.

export type ProductFieldStatus = 'detected' | 'not_printed' | 'not_visible' | 'uncertain';
export type ProductFieldSource = 'ocr' | 'vision' | 'merged' | 'user' | 'none' | 'gemini_vision';

export interface ProductField {
  value: string | null;
  status: ProductFieldStatus;
  confidence: number;
  source: ProductFieldSource;
  conflicts?: Array<{ source: string; value: string | null }> | null;
}

export type ProductInformation = Record<string, ProductField>;

// The 26 canonical fields in display order (must match PRODUCT_FIELDS).
export const PRODUCT_FIELDS: ReadonlyArray<{ key: string; label: string }> = [
  { key: 'brand_or_commodity_name', label: 'Brand / Commodity name' },
  { key: 'generic_name', label: 'Generic name' },
  { key: 'net_quantity', label: 'Net quantity' },
  { key: 'quantity_unit', label: 'Quantity unit' },
  { key: 'manufacturer_name', label: 'Manufacturer name' },
  { key: 'manufacturer_address', label: 'Manufacturer address' },
  { key: 'packer_name', label: 'Packer name' },
  { key: 'packer_address', label: 'Packer address' },
  { key: 'marketer_name', label: 'Marketer name' },
  { key: 'marketer_address', label: 'Marketer address' },
  { key: 'mrp', label: 'MRP (max retail price)' },
  { key: 'mrp_tax_inclusive', label: 'MRP tax inclusive' },
  { key: 'unit_sale_price', label: 'Unit sale price' },
  { key: 'packing_date', label: 'Packing date' },
  { key: 'manufacturing_date', label: 'Manufacturing date' },
  { key: 'expiry_date', label: 'Expiry date' },
  { key: 'batch_number', label: 'Batch number' },
  { key: 'customer_care_name', label: 'Customer care name' },
  { key: 'customer_care_phone', label: 'Customer care phone' },
  { key: 'toll_free_number', label: 'Toll-free contact' },
  { key: 'customer_care_email', label: 'Customer care email' },
  { key: 'country_of_origin', label: 'Country of origin' },
  { key: 'vegetarian_mark', label: 'Vegetarian mark' },
  { key: 'non_vegetarian_mark', label: 'Non-vegetarian mark' },
  { key: 'fssai_number', label: 'FSSAI number' },
  { key: 'certifications', label: 'Certifications' },
];

// Fields whose OCR/vision disagreement must not be silently resolved.
export const CONFLICT_SENSITIVE_FIELDS = new Set([
  'mrp',
  'net_quantity',
  'packing_date',
  'manufacturing_date',
  'expiry_date',
  'fssai_number',
]);

export function isProductInformation(value: unknown): value is ProductInformation {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return false;
  return Object.values(value).every(
    (entry) =>
      !!entry &&
      typeof entry === 'object' &&
      'status' in entry &&
      'confidence' in entry &&
      'source' in entry,
  );
}

// Legacy rows may store the JSONB payload as a JSON string.
export function parseProductInformation(raw: unknown): ProductInformation {
  if (!raw) return {};
  let value: unknown = raw;
  if (typeof raw === 'string') {
    try {
      value = JSON.parse(raw);
    } catch {
      return {};
    }
  }
  return isProductInformation(value) ? value : {};
}